README.rst

```diff
@@ -0,0 +1,112 @@
+
+.. Add a brief (few sentence) description of what this package provides.
+
+The Observatory Operations Data Service watches for files in one or more directories, and then ingests them into an LSST Butler repository.   Files are expired from the repository at specified intervals.
+
+Usage:  oods.py [[-c config][-y config] [-v]
+
+Options:
+
+.. code:: bash
+
+ -c config use specified OODS YAML configuration file
+ -y config validate YAML configuration file
+ -v give verbose output
+
+Set up and usage
+----------------
+
+1) create the Gen2 Butler repository:
+
+.. code:: bash
+
+ mkdir repo
+ echo "lsst.obs.lsst.auxTel.AuxTelMapper" > repo/_mapper
+
+2) Edit the YAML configuration file.  The default example is located in:
+
+.. code:: bash
+
+    $CTRL_OODS_DIR/etc/oods.yaml
+
+3) Run the OODS:
+
+.. code:: bash
+
+    $CTRL_OODS_DIR/oods.py -c oods.yaml
+
+NOTE:  if you run the OODS without modifying the directory paths, it expects to scan for files in the directory in which the OODS has been invoked.   It will scan the directory "data" and use the Gen2 Butler repository "repo" by default.
+
+Configuration
+-------------
+
+The OODS is configured with the following YAML file:
+
+.. code:: yaml
+
+ defaultInterval: &interval
+     days: 0
+     hours: 0
+     minutes: 0
+     seconds: 0
+ 
+ ingester:
+     directories:
+         - data
+     butler:
+         class:
+             import : lsst.ctrl.oods.gen2ButlerIngester
+             name : Gen2ButlerIngester
+         repoDirectory : repo
+     batchSize: 20
+     scanInterval:
+         <<: \*interval
+         seconds: 10
+ 
+ cacheCleaner:
+     directories:
+         - repo/raw
+     scanInterval:
+         <<: \*interval
+         seconds: 30
+     filesOlderThan:
+         <<: \*interval
+         days: 30
+     directoriesEmptyForMoreThan:
+         <<: \*interval
+         days: 1
+
+The "defaultInterval" block is used as shorthand for the intervals used throughout the rest of the YAML configuration.
+
+The "ingester" block
+--------------------
+
+This has four sections:  directories, butler, scanInterval, and batchSize.
+
+The "directories" section takes a list of directories to watch.   By default this watches the "data" directory which is expected to be in the same directory in which the OODS runs.
+
+The "butler" section specified which type of LSST Butler to run, and the repository to use.  By default, it uses an object called Gen2ButlerIngester, specified by the import "lsst.ctrl.oods.gen2ButlerIngester".   If you write your own ingestion object, follow the pattern specified in this file.   By default the "repo" directory is expected to be in the same directory which the OODS runs.  This butler repository is expected to be set up properly (see below) before the first invocation of the OODS.
+
+The "scanInterval" section specifies the frequency at which to scan the "directories" specified above.  In the example, it scans every 10 seconds.
```

Comment:
The API to do that, inotify(), does that on a per directory basis, and is what I originally intended to use.   There is no python extension for this system call API in Python 3.  However, I wrote a pure Python extension myself for Linux, and which is in the lsst-dm/ctrl_notify package.  This has unit tests, but has not been reviewed.

The original suggestion for how the OODS would handle files was having them come into one directory.    While the discussions of various parts were being discussed, I wrote that lsst-dm/ctrl_notify package.   Further discussions concluded that because of the number of files involved, a tree of directories would be required so that no single directory would become too large and slow down performance.  (That was an issues in other projects when all files were dumped into one directory and lookups, and even directory listings too a long time).

The Linux inotify() system call can only handle one directory at a time.   In order to watch new directories as they're detected, those directories must be added to the inotify() watch list.   There's a race condition that exists at that point.  It's possible to create a new directory and write files below it, before the inotify() interface reports the new directory exists.  Once the new directory is registered, only new files (or modifications) are reported;  the files that already existed are not.

Removing registration from directories that exist, but have been abandoned would also be required, and inotify() does not support that.  That can only be done via scanning.  (And now that I bring this up, I realize the empty directory removal process I use for the repo needs to be applied to the incoming directory as well, unless we plan on having the writers manage that.  I'll mark that down as a TODO item).  We would need to do directory removal because there is a (settable) upper limit to the number of inotify() watch points that occur, and if this limit is reached, read failures occur.

It may be possible to have a combination of kernel notifications and scanning (ie, register the directory, scan for the files, process those files, don't scan again since files from now on will be giving via inotify), but I haven't investigated that.  There may be other race conditions I haven't considered.

Reply:
I believe it would.   However, there are several issues that could arise.  First, it was my intent to try and have the OODS be unaware of the directory structure of the incoming files.  This would keep the writing side and the OODS side somewhat independent.  If this changed, and the OODS created the directories, that would need to be coordinated.  Second, it was also the intent that file writing to the incoming disk directory could happen while the OODS was offline;  when the OODS comes up back up, it immediately checks to see what files might have been written while it was "gone".  Third, If the OODS is offline, when it came back up it would first need to create any directories that should have been created during that downtime, before letting anything write files to the incoming directory to avoid race conditions.

If this is really just tens of thousands of images per day at worst, is the current mechanism bad?  What would be the "win" of the kernel notify method, given the extra steps we would have to take?

Reply:
Ah, yes, catch-up does pose some issues.  This obviously needs some thoughtful design rather than insta-comments by a kibitzer.

The "win" of kernel notify is shorter latency, which is a prime goal of the OODS.